doc/subsystems/usb/usb.rst

```diff
@@ -52,245 +52,57 @@ USB device class drivers
-      COMMUNICATION_DEVICE_CLASS_DATA,/* Class */
-      0x00,                           /* SubClass */
-      0x00,                           /* Protocol */
-      0x00,                           /* Index of the Interface String Descriptor */
-
-      /* First Endpoint IN */
-      USB_ENDPOINT_DESC_SIZE,         /* Descriptor size */
-      USB_ENDPOINT_DESC,              /* Descriptor type */
-      CDC_ENDP_IN,                    /* Endpoint address */
-      USB_DC_EP_BULK,                 /* Attributes */
-      LOW_BYTE(CDC_BULK_EP_MPS),
-      HIGH_BYTE(CDC_BULK_EP_MPS),     /* Max packet size */
-      0x00,                           /* Interval */
-
-      /* Second Endpoint OUT */
-      USB_ENDPOINT_DESC_SIZE,         /* Descriptor size */
-      USB_ENDPOINT_DESC,              /* Descriptor type */
-      CDC_ENDP_OUT,                   /* Endpoint address */
-      USB_DC_EP_BULK,                 /* Attributes */
-      LOW_BYTE(CDC_BULK_EP_MPS),
-      HIGH_BYTE(CDC_BULK_EP_MPS),     /* Max packet size */
-      0x00,                           /* Interval */
-
-      /* String descriptor language, only one, so min size 4 bytes.
-       * 0x0409 English(US) language code used
-       */
-      USB_STRING_DESC_SIZE,           /* Descriptor size */
-      USB_STRING_DESC,                /* Descriptor type */
-      0x09,
-      0x04,
-      /* Manufacturer String Descriptor "Intel" */
-      0x0C,
-      USB_STRING_DESC,
-      'I', 0, 'n', 0, 't', 0, 'e', 0, 'l', 0,
-      /* Product String Descriptor "CDC-ACM" */
-      0x10,
-      USB_STRING_DESC,
-      'C', 0, 'D', 0, 'C', 0, '-', 0, 'A', 0, 'C', 0, 'M', 0,
-      /* Serial Number String Descriptor "00.01" */
-      0x0C,
-      USB_STRING_DESC,
-      '0', 0, '0', 0, '.', 0, '0', 0, '1', 0,
-   };
-
-.. code-block:: c
-
-   static struct usb_ep_cfg_data cdc_acm_ep_data[] = {
-      {
-         .ep_cb = cdc_acm_int_in,
-         .ep_addr = CDC_ENDP_INT
-      },
-      {
-         .ep_cb = cdc_acm_bulk_out,
-         .ep_addr = CDC_ENDP_OUT
-      },
-      {
-         .ep_cb = cdc_acm_bulk_in,
-         .ep_addr = CDC_ENDP_IN
-      }
-   };
-
-.. code-block:: c
-
-   static struct usb_cfg_data cdc_acm_config = {
-      .usb_device_description = cdc_acm_usb_description,
-      .cb_usb_status = cdc_acm_dev_status_cb,
-      .interface = {
-      .class_handler = cdc_acm_class_handle_req,
-      .custom_handler = NULL,
-      .payload_data = NULL,
-      },
-      .num_endpoints = CDC1_NUM_EP + CDC2_NUM_EP,
-      .endpoint = cdc_acm_ep_data
-   };
-
-.. code-block:: c
-
-   ret = usb_set_config(&cdc_acm_config);
-   if (ret < 0) {
-      DBG("Failed to config USB\n");
-      return ret;
-   }
+should call :c:func:`usb_set_config()` passing as parameter the instance's
+configuration structure.
 
-To enable the USB device for host/device connection:
+For example, for USB loopback application:
+
+.. literalinclude:: ../../../subsys/usb/class/loopback.c
```

Comment:
There is some danger using literal includes by line number of the actual code - there's no check that the lines specified are what was intended. If someone were to add or delete a few lines to the source code, we'll pull in the wrong lines for the documentation.  Perhaps an obvious comment in ``loopback.c`` saying code here is included in doc/subsystems/usb/usb.rst  would help, but I'm more inclined to directly copy the code here into the documentation since it's a sample.

@galak, your thoughts?

Reply:
``\snippet`` is used to include source code within the doxygen-generated content.  The equivalent for our Sphinx documentation is to use the ``:start-after:`` and ``:end-before:`` options on the ``literal-include:`` directive.  You'd need to add unique comments in the source code to delimit the start and stop regions in the source.  For example, in loopback.c you could add comments before and after lines 32-71:

```
/* usb.rst config structure start */
struct usb_loopback_config {
	struct usb_if_descriptor if0;
	struct usb_ep_descriptor if0_out_ep;
	struct usb_ep_descriptor if0_in_ep;
} __packed;

. . .

		.bmAttributes = USB_DC_EP_BULK,
		.wMaxPacketSize = sys_cpu_to_le16(CONFIG_LOOPBACK_BULK_EP_MPS),
		.bInterval = 0x00,
	},
};
/* usb.rst config structure end */

```

and then in usb.rst use:

```
.. literalinclude:: ../../../subsys/usb/class/loopback.c
   :language: c
   :start-after: usb.rst config structure start
   :end-before: usb.rst config structure end
   :linenos:
```